### PR TITLE
Add Iterator#apply Iterator#chainApply to improve iterator call chaining

### PR DESCRIPTION
**Pull request by lundibundi**
- [x] code is properly formatted (`npm run fmt`)
- [x] tests are added/updated
- [x] documentation is updated (`npm run doc` to regenerate documentation based on comments)
- [x] description of changes is added under the `Unreleased` header in CHANGELOG.md

### CODE REVIEW

File: lib/iterator.js [hunk 1]
```diff
@@ -199,6 +199,23 @@ class Iterator {
...
+  }
+
+  // Call a function with `this` and wrap the result in an Iterator.
+  // Example:
+  // iter([1, 2])
+  //   .chainApply(([a, b]) => [a + b, a - b])
+  //   .flat()
+  //   .join(', ');
+  //  => '3, -1'
+  chainApply(fn) {
```
Comment (SemenchenkoVitaliy) [inline lib/iterator.js]:
```suggestion
  // Returns: <Iterator>
  chainApply(fn) {
```

File: lib/iterator.js [hunk 2]
```diff
@@ -199,6 +199,23 @@ class Iterator {
...
+    return fn(this);
+  }
+
+  // Call a function with `this` and wrap the result in an Iterator.
+  // Example:
+  // iter([1, 2])
+  //   .chainApply(([a, b]) => [a + b, a - b])
+  //   .flat()
+  //   .join(', ');
+  //  => '3, -1'
```
Comment (SemenchenkoVitaliy) [inline lib/iterator.js]:
```suggestion
  // Result: '3, -1'
```

File: lib/iterator.js [hunk 3]
```diff
@@ -199,6 +199,23 @@ class Iterator {
...
+  //   fn <Function>
+  apply(fn) {
+    return fn(this);
+  }
+
+  // Call a function with `this` and wrap the result in an Iterator.
+  // Example:
+  // iter([1, 2])
+  //   .chainApply(([a, b]) => [a + b, a - b])
+  //   .flat()
```
Comment (SemenchenkoVitaliy) [inline lib/iterator.js]: I don't like that you'll need to use `flat()` in such use case.

Comment (lundibundi) [reply lib/iterator.js]: I considered always calling `iter(fn(this))` but that wouldn't work nicely with `string` use-case.

File: lib/iterator.js [hunk 4]
```diff
@@ -199,6 +199,28 @@ class Iterator {
     return result + suffix;
   }
 
+  // Call a function with `this`. Will be equivalent to calling `fn(it)`.
+  //   fn <Function>
+  //     this <Iterator>
```
Comment (nechaido) [inline lib/iterator.js]: IMHO we should specify that `apply` returns result of applying `fn`.

File: CHANGELOG.md [hunk 5]
```diff
@@ -8,6 +8,11 @@ and this project adheres to
 
 ## [Unreleased][unreleased]
 
+## Added
+
+- `Iterator#apply` and `Iterator#chainApply` to improve iterator
```
Comment (belochub) [inline CHANGELOG.md]:
```suggestion
- `Iterator#apply()` and `Iterator#chainApply()` to improve iterator
```

**Comment by belochub** [comment]
LGTM. Landed in https://github.com/metarhia/common/commit/a1a2dccf0f950714f7dfc93c899cd8862dc5dce3.

**Pull request closed without merging**